cmem/encoder.go

```diff
@@ -0,0 +1,171 @@
+	Buf    []byte
+	Offset int
+}
+
+// Encode encodes the value passed as data to binary form stored in []Buf. This
+// buffer is a Go representation of a C in-memory object that can be appended
+// to e.g. a HDF5 PacketTable.
+//
+// Struct values must only have exported fields, otherwise Encode will panic.
+func (enc *Encoder) Encode(data interface{}) error {
+	padding := enc.Offset - len(enc.Buf)
+
+	if padding > 0 {
+		enc.Buf = append(enc.Buf, make([]byte, padding)...)
+	}
+
+	if data, ok := data.(CMarshaler); ok {
+		raw, err := data.MarshalC()
+		if err != nil {
+			return err
+		}
+
+		enc.Buf = append(enc.Buf, raw...)
+
+		return nil
+	}
+
+	rv := reflect.Indirect(reflect.ValueOf(data))
+	rt := rv.Type()
+
+	switch rt.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < rv.Len(); i++ {
+			if err := enc.Encode(rv.Index(i).Interface()); err != nil {
+				return err
+			}
+		}
+		return nil
+
+	case reflect.Struct:
+		offset := enc.Offset
+		for i := 0; i < rv.NumField(); i++ {
+			sfv := rv.Field(i).Interface()
+			// In order to keep the struct offset always correct
+			enc.Offset = offset + int(rt.Field(i).Offset)
+			if err := enc.Encode(sfv); err != nil {
+				return err
+			}
+			// Reset the offset to the correct array sized
+			enc.Offset = offset + int(rt.Size())
+		}
+		return nil
+
+	case reflect.String:
+		str := append([]byte(rv.String()), 0)
+		strPtr := uintptr(unsafe.Pointer(&str[0]))
+		dataSize := int(unsafe.Sizeof(strPtr))
+
+		switch dataSize {
+		case 4:
+			var raw [4]byte
+			nativeEndian.PutUint32(raw[:], uint32(strPtr))
+			enc.Buf = append(enc.Buf, raw[:]...)
+		case 8:
+			var raw [8]byte
+			nativeEndian.PutUint64(raw[:], uint64(strPtr))
+			enc.Buf = append(enc.Buf, raw[:]...)
+		}
+
+		enc.Offset += dataSize
+
+	case reflect.Ptr:
+		return enc.Encode(rv.Elem())
+
+	case reflect.Int8:
+		enc.Buf = append(enc.Buf, byte(rv.Int()))
+		enc.Offset++
+
+	case reflect.Uint8:
+		enc.Buf = append(enc.Buf, byte(rv.Uint()))
+		enc.Offset++
+
+	case reflect.Int16:
+		var raw [2]byte
+		nativeEndian.PutUint16(raw[:], uint16(rv.Int()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 2
+
+	case reflect.Uint16:
+		var raw [2]byte
+		nativeEndian.PutUint16(raw[:], uint16(rv.Uint()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 2
+
+	case reflect.Int32:
+		var raw [4]byte
+		nativeEndian.PutUint32(raw[:], uint32(rv.Int()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 4
+
+	case reflect.Uint32:
+		var raw [4]byte
+		nativeEndian.PutUint32(raw[:], uint32(rv.Uint()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 4
+
+	case reflect.Int64:
+		var raw [8]byte
+		nativeEndian.PutUint64(raw[:], uint64(rv.Int()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 8
+
+	case reflect.Uint64:
+		var raw [8]byte
+		nativeEndian.PutUint64(raw[:], uint64(rv.Uint()))
+		enc.Buf = append(enc.Buf, raw[:]...)
+		enc.Offset += 8
+
+	case reflect.Float32:
+		enc.Offset += 4
```

Comment:
Please drop this to below the append to match the other stanzas.

Reply:
Good point.